src/ssl_bn.c

```diff
@@ -2366,61 +2366,84 @@ int wolfSSL_BN_print_fp(XFILE fp, const WOLFSSL_BIGNUM *bn)
  * BN_CTX APIs
  ******************************************************************************/
 
-/* Allocate and return a new BN context object.
+/* Create a new BN context object.
  *
- * BN context not needed for operations.
- *
- * @return  Pointer to dummy object.
+ * @return  BN context object on success.
+ * @return  NULL on failure.
  */
 WOLFSSL_BN_CTX* wolfSSL_BN_CTX_new(void)
 {
-    /* wolfcrypt doesn't need BN context. */
-    static int ctx;
+    WOLFSSL_BN_CTX* ctx = NULL;
+
     WOLFSSL_ENTER("wolfSSL_BN_CTX_new");
-    return (WOLFSSL_BN_CTX*)&ctx;
+    ctx = (WOLFSSL_BN_CTX*)XMALLOC(sizeof(WOLFSSL_BN_CTX), NULL,
+            DYNAMIC_TYPE_OPENSSL);
+    if (ctx != NULL) {
+        wolfSSL_BN_CTX_init(ctx);
+    }
+
+    return ctx;
 }
 
 /* Initialize a BN context object.
  *
- * BN context not needed for operations.
- *
- * @param [in] ctx  Dummy BN context.
+ * @param [in] ctx  BN context object.
  */
 void wolfSSL_BN_CTX_init(WOLFSSL_BN_CTX* ctx)
 {
-    (void)ctx;
     WOLFSSL_ENTER("wolfSSL_BN_CTX_init");
+    if (ctx != NULL) {
+        XMEMSET(ctx, 0, sizeof(WOLFSSL_BN_CTX));
```

Comment:
Old implementations in OpenSSL reset the pool and isn't meant to be called in BN_CTX_new().
API only in 1.0.2 and earlier.

Reply:
Its called in the original implementations of `BN_CTX_new` (https://github.com/openssl/openssl/commit/9b141126d4b6f0636bc047e81b846c193ae26611#diff-9ef8217a3064ad4a970c33b34c9c8cabc4d8452ae282e5c1a19cbb0d8793fd88R74). Would you like me to separate the functions out? Maybe we should deprecate `wolfSSL_BN_CTX_init` as its equivalent is no longer in OpenSSL for a long time.

Reply:
*sigh* our implementation is not doing the right thing then.
BN_CTX_init is an OpenSSL API - our implementation has to match theirs.
Definitely deprecate it!